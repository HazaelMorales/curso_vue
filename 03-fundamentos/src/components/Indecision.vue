<template>
    <div>
        <h1>Indecision</h1>
        <!--Valida si hay una imagen para cargarla-->
        <img v-if="img" :src="img" alt="bg"> 
        <div class="bg-dark"></div> 

        <div class="indecision-container">
            <input 
                type="text" 
                placeholder="Hazme una pregunta"
                v-model="question">

            <p>Recuerda terminar con un signo de interrogacion (?)</p>

            <!--Valida si la pregunta es valida (true), mostrara la pregunta y la respuesta-->
            <div v-if="isValidQuestion">
                <h2>{{ question }}</h2>
                <h1>{{ answer }}</h1>
            </div> 
        </div>  
    </div>
</template>

<script>
export default {
    /*
        Función que retorna los datos locales del componente
        
        ARGS:
            question -> nulo
            answer -> nulo
            img -> nulo
            isValidQuestion -> falso
    */
    data(){
        return {
            question: null,
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },

    methods: {
    // Método para obtener una respuesta de la API
    async getAnswer() {
            try{
                // Se establece la respuesta inicial mientras se espera la respuesta de la API
                this.answer = 'Pensando....'
                
                // Se hace una solicitud a la API y se espera la respuesta
                const { answer, image } = await fetch('https://yesno.wtf/api').then(response => response.json())
                
                // Se actualiza la respuesta con el resultado obtenido de la API
                this.answer = answer === 'yes' ? 'Si!' : 'No'
                
                // Se actualiza la imagen con el resultado obtenido de la API
                this.img = image
            
            }catch(error){
                console.log('IndecisionComponent: ', error)
                this.answer = 'No se pudo cargar el API'
                this.img = null
            }
        }
    },

    watch: {
        // Vigilancia de cambios en la pregunta
        question(value, oldValue) {
            // Se establece inicialmente que la pregunta no es válida
            this.isValidQuestion = false

            console.log({value})
            
            // Si la pregunta no incluye un signo de interrogación, no se realiza nada más
            if (!value.includes('?')) return
            
            // Si la pregunta es válida, se actualiza el estado de isValidQuestion y se llama al método getAnswer()
            this.isValidQuestion = true
            this.getAnswer()
        }
    }
}
</script>

<style scoped>
    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }
</style>