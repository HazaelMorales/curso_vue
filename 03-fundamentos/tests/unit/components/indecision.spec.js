//Importamos unas herrmientas de test de vue
import {shallowMount, mount} from '@vue/test-utils'

//Importamos el componente
import Indecision from '@/components/Indecision'

describe("Indecision Component",()=>{
    let wrapper
    let consoleLogSpy
    
    //CREACION DE UN MOCK
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    // beforeEach() se ejecuta antes de cada prueba, y aquí se inicializa el wrapper
    beforeEach(() => {
        // shallowMount se utiliza para montar el componente Counter sin renderizar sus componentes hijos
        wrapper = shallowMount(Indecision)
        
        // Se crea un espía para el método console.log
        consoleLogSpy = jest.spyOn(console, 'log')

        // Se limpian todos los mocks creados
        jest.clearAllMocks()
    })

    test('Debe de hacer match con el snapshot', () => {
        // Se compara el HTML renderizado del componente con el snapshot existente
        expect(wrapper.html()).toMatchSnapshot()
    })
    
    test('Escribir en el input no debe de disparar nada (console.log)', async () => {
        // Se espía el método getAnswer del componente
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    
        // Se encuentra el elemento de input y se establece su valor
        const input = wrapper.find('input')
        await input.setValue('Hola Mundo')
    
        // Se verifica que console.log no haya sido llamado y que getAnswer no haya sido llamado
        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()
    })
    
    test('Escribir el símbolo de "?" debe de disparar el fetch', async () => {
        // Se espía el método getAnswer del componente
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    
        // Se encuentra el elemento de input y se establece su valor
        const input = wrapper.find('input')
        await input.setValue('Hola Mundo?')
    
        // Se verifica que console.log haya sido llamado una vez y que getAnswer haya sido llamado
        expect(consoleLogSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalled()
    })
    
    test('Pruebas en getAnswer', async () => {
        // Se llama al método getAnswer del componente
        await wrapper.vm.getAnswer()
    
        // Se encuentra el elemento img
        const img = wrapper.find('img')
    
        // Se verifica que el elemento img exista, que la propiedad img del componente sea la esperada y que la propiedad answer sea la esperada
        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Si!')
    })
    
    test('Pruebas en getAnwser - Fallo en el API', async () => {
        // Se simula un fallo en la llamada al API
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))
    
        // Se llama al método getAnswer del componente
        await wrapper.vm.getAnswer()
    
        // Se encuentra el elemento img
        const img = wrapper.find('img')
    
        // Se verifica que el elemento img no exista y que la propiedad answer del componente sea la esperada
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo cargar el API')
    })
})