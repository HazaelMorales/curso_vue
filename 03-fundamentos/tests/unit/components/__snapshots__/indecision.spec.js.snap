// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>Indecision</h1>
  <!--Valida si hay una imagen para cargarla-->
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
    <p>Recuerda terminar con un signo de interrogacion (?)</p>
    <!--Valida si la pregunta es valida (true), mostrara la pregunta y la respuesta-->
    <!--v-if-->
  </div>
</div>
`;
